src/plugins/syscalls/syscalls.cpp

```diff
@@ -107,283 +107,92 @@
-    else
-        printf("0x%" PRIx64 ",", static_cast<uint64_t>(val));
-
-    if ( str )
-    {
-        printf("%s", str);
-    }
-
-    printf(",");
-}
-
-static void print_kv_arg(syscalls* s, drakvuf_t drakvuf, drakvuf_trap_info_t* info, const arg_t& arg, addr_t val, const char* str)
-{
-    if ( str )
-    {
-        printf(",%s=\"%s\"", arg.name, str);
-        return;
-    }
-
-    if ( 4 == s->reg_size )
-        printf(",%s=0x%" PRIx32, arg.name, static_cast<uint32_t>(val));
-    else
-        printf(",%s=0x%" PRIx64, arg.name, static_cast<uint64_t>(val));
-}
-
-
-static void print_json_arg(syscalls* s, drakvuf_t drakvuf, drakvuf_trap_info_t* info, const syscall_t* sc, size_t i, addr_t val, const char* str)
-{
-    const arg_t& arg = sc->args[i];
-
-    if ( str )
-    {
-        gchar* escaped = drakvuf_escape_str(str);
-        printf("{\"%s\" : %s}", arg.name, escaped);
-        g_free(escaped);
-    }
-    else
-    {
-        if ( 4 == s->reg_size )
-            printf("{\"%s\" :%"  PRIu32 "}", arg.name, static_cast<uint32_t>(val));
-        else
-            printf("{\"%s\" :%" PRIu64 "}", arg.name, static_cast<uint64_t>(val));
-    }
-
-    if (i < sc->num_args-1)
-        printf(",");
-}
-
-static void print_default_arg(syscalls* s, drakvuf_t drakvuf, drakvuf_trap_info_t* info, const arg_t& arg, addr_t val, const char* str)
-{
-    printf("\t%s %s %s: ", arg_direction_names[arg.dir], type_names[arg.type], arg.name);
-
-    if ( 4 == s->reg_size )
-        printf("0x%" PRIx32, static_cast<uint32_t>(val));
-    else
-        printf("0x%" PRIx64, static_cast<uint64_t>(val));
-
-    if ( str )
+    std::string str;
+    if (cstr)
     {
-        printf(" -> '%s'", str);
+        str = std::string(cstr);
+        g_free(cstr);
     }
-
-    printf("\n");
+    return str;
 }
 
-void print_args(syscalls* s, drakvuf_t drakvuf, drakvuf_trap_info_t* info, const syscall_t* sc, void* args_data)
+void print_syscall(syscalls* s, drakvuf_t drakvuf, os_t os,
+                   bool syscall, drakvuf_trap_info_t* info,
+                   int nr, const char *module, const syscall_t *sc,
+                   unsigned int nargs, std::vector<uint64_t>& args,
+                   uint64_t ret, const char *extra_info)
 {
-    if ( !args_data )
-        return;
-
-    size_t nargs = sc->num_args;
-    uint32_t* args_data32 = (uint32_t*)args_data;
-    uint64_t* args_data64 = (uint64_t*)args_data;
+    if (sc)
+        info->trap->name = sc->name;
 
-    for ( size_t i=0; i<nargs; i++ )
+    if (syscall)
     {
-        addr_t val = 0;
-
-        if ( 4 == s->reg_size )
-            memcpy(&val, &args_data32[i], sizeof(uint32_t));
-        else
-            memcpy(&val, &args_data64[i], sizeof(uint64_t));
-
-        char* str = extract_string(s, drakvuf, info, sc->args[i], val);
-
-        switch (s->format)
+        std::vector<std::pair<const char*, std::variant<fmt::Qstr<std::string>, fmt::Xval<unsigned long>>>> fmt_args;
```

Comment:
This makes my head hurt.

Reply:
My too :) I don't like templates. But without this KV output is broken. I don't know how to re-implement it with current template-based API.

Reply:
I could make something like this:
```c++
    if (syscall)
    {
        std::vector<std::pair<const char*, fmt::Qstr<std::string>>> str_args;
        std::vector<std::pair<const char*, fmt::Xval<uint64_t>>> xval_args;
        for (size_t i = 0; i < args.size(); ++i)
        {
            auto str = extract_string(s, drakvuf, info, sc->args[i], args[i]);
            if ( !str.empty() )
                str_args.push_back(keyval(sc->args[i].name, fmt::Qstr(str)));
            else
                xval_args.push_back(keyval(sc->args[i].name, fmt::Xval(args[i])));
        }

        fmt::print(s->format, "syscall", drakvuf, info,
            keyval("Module", fmt::Qstr(module)),
            keyval("vCPU", fmt::Nval(info->vcpu)),
            keyval("CR3", fmt::Xval(info->regs->cr3)),
            keyval("Syscall", fmt::Nval(nr)),
            keyval("NArgs", fmt::Nval(args.size())),
            str_args,
            xval_args
        );
```

But this would break CSV format.

Reply:
I will try to add new type into `output_format/common.h`

Reply:
@tklengyel, would it be acceptable to add(?):
```c++
std::vector<std::pair<std::string, std::any>> fmt_args;
```

Reply:
It's very beautiful tho :(

But ok, if we want to change this, then I believe that `std::any` is too broad here, we can create a new type using `std::variant` with all the types from `fmt` and call it `fmt_arg` or smth.
```cpp
using fmt_arg = std::variant<fmt::Qstr<std::string>, fmt::Xval<unsigned long>>
```

Reply:
@BonusPlay, thank you very much!

> why const char* when you have std::string

That is an obsolete trash from multiple re-works and re-bases. Thank you to notice!

Reply:
@BonusPlay, fixed.

@tklengyel is it acceptable?